jdaviz/app.py

```diff
@@ -542,6 +543,51 @@ def get_subsets_from_viewer(self, viewer_reference, data_label=None, subset_type
                                          cls=None)
         regions = {}
 
+        def _get_all_subregions(mask, spec_axis_data):
+            """
+            Return all subregions within a subset.
+
+            Parameters
+            ----------
+            mask : list
+                List of indices in spec_axis_data that are part of the subset.
+            spec_axis_data : list
+                List of spectral axis values.
+            Returns
+            -------
+            combined_spec_region : `~specutils.SpectralRegion`
+                SpectralRegion object containing all subregions of the subset.
+            """
+            if len(mask) == 0:
+                # Mask should only be 0 if ApplyROI is used to incorrectly
+                # create subsets via the API
+                raise ValueError
+
+            current_edge = 0
+            combined_spec_region = None
+            for index in range(1, len(mask)):
+                # Find spot where mask == True is for a different region of the subset
+                # i.e. mask = [0, 1, 4, 5]
+                # mask[2] != mask[1] + 1
+                if mask[index] != mask[index - 1] + 1:
+                    subset_region = spec_axis_data[mask[current_edge]: mask[index - 1] + 1]
+                    if not combined_spec_region:
+                        combined_spec_region = SpectralRegion(min(subset_region),
+                                                              max(subset_region))
+                    else:
+                        combined_spec_region += SpectralRegion(min(subset_region),
+                                                               max(subset_region))
+                    current_edge = index
+
+            # Get last region within the subset
+            if current_edge != index:
+                subset_region = spec_axis_data[mask[current_edge]: mask[index]]
+                # No if check here because len(mask) must be greater than 1
+                # so combined_spec_region will have been instantiated in the for loop
+                combined_spec_region += SpectralRegion(min(subset_region), max(subset_region))
```

Comment:
I managed to find a situation where this is attempting to add the region to None, trying to figure out how that was triggered in the code. Through the UI, I first made a Subset 1 with a few subregions, then created a Subset 2 with a single region. After that, I hit the `TypeError` that this says shouldn't be possible to hit.

Reply:
Ok, I found the cause here. If you make a single sub-region subset with the `replace` function, you get something different than if you do so with the `add` function. It looks the same on the front end, but Glue represents them differently and they end up looking different when returned by our `get_subsets_from_viewer` function. In the sample below, Subset 2 used `add` and Subset 3 used `replace`:

```
>>> s2.subset_state
<glue.core.subset.OrState at 0x7f9539e12970>

>>>s3.subset_state
<glue.core.subset.RangeSubsetState at 0x7f9548b09490>
```

And our app method returns:

```
'Subset 2': Spectral Region, 1 sub-regions:
   (5499.20556640625 Angstrom, 5897.93505859375 Angstrom) ,
 'Subset 3': <RectanglePixelRegion(center=PixCoord(x=7616.167200488471, y=0), width=418.5868159615429, height=0, angle=0.0 deg)>
```

It seems like we should return the same thing for a single-region Subset, no matter which method was used to create it, which I guess will require working around the Glue backend differences. This might be a good time to tag in @astrofrog and ask if something changed on the backend for Glue subsets recently, since as we've said before I'm 99% this had been working at some point in the past. (@astrofrog I know you're busy with astropy release right now, this is just a heads up for when you have time).

Reply:
Ah, wait, I realized why the model fitting used to work with subsets with subregions - we used to just grab e.g. `Subset 1` as the _data_, whereas now we've split out the data and spectral subsets into separate menus. So previously it was just grabbing the subset as a masked `Spectrum1D`, which worked fine for fitting. There were probably always problems with the subset retrieval for multiple subregions and we just didn't notice until we switched to trying to apply the subset separately for the model fitting.

Reply:
I would kind of expect any of our subset getters would resolve all regions into `specutils.spectra.SpectralRegion` objects at the end of the day, so I'm surprised we're getting a PixelRegion out of it.